src/cudamatrix/cu-matrix.cc

```diff
@@ -2540,33 +2577,54 @@ void CuMatrixBase<Real>::Lookup(const std::vector<Int32Pair> &indices,
   // Checks the pointer.
   KALDI_ASSERT(output != NULL);
 
-  // Resizes the output vector.
-  output->resize(indices.size());
+#if HAVE_CUDA == 1
+  CuArray<Int32Pair> cuda_indices(indices);
+  Lookup(cuda_indices, output);
+#else
+  for (int32 i = 0; i < indices.size(); i++) {
+    output[i] = (*this)(indices[i].first, indices[i].second);
+  }
+#endif
+}
+
+template<typename Real>
+void CuMatrixBase<Real>::Lookup(const CuArray<Int32Pair> &indices,
+				Real *output) const {
+  int32 num_elements = indices.Dim();
+  // Checks the dimension.
+  MatrixIndexT num_rows = this->num_rows_, num_cols = this->num_cols_;
+  std::vector<Int32Pair> cpu_indices;
+  indices.CopyToVec(&cpu_indices);
+  for (int32 i = 0; i < num_elements; ++i) {
+    KALDI_ASSERT(cpu_indices[i].first < num_rows && cpu_indices[i].first >= 0 &&
+                 cpu_indices[i].second < num_cols && cpu_indices[i].second >= 0);
+  }
+  
+  // Checks the pointer.
+  KALDI_ASSERT(output != NULL);
 
 #if HAVE_CUDA == 1
   if (CuDevice::Instantiate().Enabled()) {
-    CuArray<Int32Pair> cuda_indices(indices);
-    CuArray<Real> cuda_output(output->size());
-
+    CuArray<Real> cuda_output(num_elements);
     Timer tim;
     dim3 dimBlock(CU1DBLOCK, 1);
-    dim3 dimGrid(n_blocks(indices.size(), CU1DBLOCK), 1);
-    
+    dim3 dimGrid(n_blocks(num_elements, CU1DBLOCK), 1);
+
     cuda_matrix_lookup(dimGrid, dimBlock, this->data_, this->Dim(),
-                       cuda_indices.Data(), indices.size(), cuda_output.Data());
+                       indices.Data(), num_elements, cuda_output.Data());
     CU_SAFE_CALL(cudaGetLastError());
-
-    cuda_output.CopyToVec(output);
     
+    cuda_output.CopyToVec(output);
     CuDevice::Instantiate().AccuProfile(__func__, tim.Elapsed());
   } else
 #endif
   {
-    for (int32 i = 0; i < indices.size(); i++) {
-      (*output)[i] = (*this)(indices[i].first, indices[i].second);
+    for (int32 i = 0; i < num_elements; i++) {
```

Comment:
You don't need to copy to cpu_indices in order to access the indices on CPU, if CUDA is not compiled in or is not being used.  In that case, the CuArray Data() pointer will point to a CPU pointer, which you can index directly.